extensions-core/protobuf-extensions/src/main/java/org/apache/druid/data/input/protobuf/ProtobufInputRowParser.java

```diff
@@ -100,16 +102,27 @@ void initDescriptor()
       parser = parseSpec.makeParser();
       initDescriptor();
     }
-    String json;
-    try {
-      DynamicMessage message = DynamicMessage.parseFrom(descriptor, ByteString.copyFrom(input));
-      json = JsonFormat.printer().print(message);
-    }
-    catch (InvalidProtocolBufferException e) {
-      throw new ParseException(e, "Protobuf message could not be parsed");
+    Map<String, Object> record;
+
+    if (parseSpec instanceof JSONParseSpec && ((JSONParseSpec) parseSpec).getFlattenSpec().getFields().isEmpty()) {
+      try {
+        DynamicMessage message = DynamicMessage.parseFrom(descriptor, ByteString.copyFrom(input));
+        record = CollectionUtils.mapKeys(message.getAllFields(), k -> k.getJsonName());
+      }
+      catch (InvalidProtocolBufferException ex) {
+        throw new ParseException(ex, "Protobuf message could not be parsed");
+      }
+    } else {
+      try {
+        DynamicMessage message = DynamicMessage.parseFrom(descriptor, ByteString.copyFrom(input));
+        String json = JsonFormat.printer().print(message);
+        record = parser.parseToMap(json);
+      }
+      catch (InvalidProtocolBufferException e) {
+        throw new ParseException(e, "Protobuf message could not be parsed");
+      }
```

Comment:
I suggest refactoring these modifications to reduce code duplications.

```suggestion
    try {
      DynamicMessage message = DynamicMessage.parseFrom(descriptor, ByteString.copyFrom(input));
    }
    catch (InvalidProtocolBufferException ex) {
      throw new ParseException(ex, "Protobuf message could not be parsed");
    }

    Map<String, Object> record;
    if (parseSpec instanceof JSONParseSpec && ((JSONParseSpec) parseSpec).getFlattenSpec().getFields().isEmpty()) {
      record = CollectionUtils.mapKeys(message.getAllFields(), k -> k.getJsonName());
    } else {
      String json = JsonFormat.printer().print(message);
      record = parser.parseToMap(json);
    }
```